File: Trabalho_docker_4/api/linkextractor.py
#!/usr/bin/env python3

import sys
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin

HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"}

def extract_links(url):
    res = requests.get(url, timeout=10, headers=HEADERS)
    res.raise_for_status()
    soup = BeautifulSoup(res.text, "html.parser")
    base = url
    links = []
    for link in soup.find_all("a"):
        href = link.get("href")
        if not href:
            continue
        links.append({
            "text": " ".join(link.text.split()) or "[IMG]",
            "href": urljoin(base, href),
        })
    return links


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("\nUsage:\n\t{} <URL>\n".format(sys.argv[0]))
        sys.exit(1)
    for link in extract_links(sys.argv[-1]):
        print("[{}]({})".format(link["text"], link["href"]))

File: Trabalho_docker_4/api/linkextractor.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
"""
API Link Extractor - Ruby com Cache Redis
Autor: Joey
"""

require 'sinatra'
require 'nokogiri'
require 'redis'
require 'json'

# Configurações
set :port, 4567
set :bind, '0.0.0.0'

# Conexão com Redis
redis = Redis.new(url: ENV['REDIS_URL'] || 'redis://localhost:6379')

# GET da raiz - instruções
get '/' do
  "Usage: http://<hostname>[:<port>]/api/<url>"
end

# GET para extrair links com cache
get '/api/*' do
  url = params['splat'][0]
  
  # Adicionar query string se existir
  query_string = request.query_string
  url += "?#{query_string}" if query_string != ""
  
  # Verificar cache
  cached = redis.get(url)
  
  if cached
    links = JSON.parse(cached)
  else
    # Extrair links
    begin
      response = HTTParty.get(url, timeout: 10)
      doc = Nokogiri::HTML(response.body)
      
      links = []
      doc.css('a').each do |link|
        href = link['href']
        # [FIX] Apenas adicionar se href existir (não nil/vazio)
        next unless href
        text = link.text.strip.split.join(' ')
        text = '[IMG]' if text.empty?
        
        links << {
          'text' => text,
          'href' => href
        }
      end
      
      # Armazenar em cache
      redis.set(url, JSON.generate(links))
    rescue => e
      puts "ERROR: #{e.class}: #{e.message}"
      links = []
    end
  end
  
  content_type :json
  JSON.pretty_generate(links)
end
